driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
+                                    return;
+                                }
+
+                                String currentKeyspace = connection.keyspace();
+                                String prepareKeyspace = toPrepare.getQueryKeyspace();
+                                if (prepareKeyspace != null && (currentKeyspace == null || !currentKeyspace.equals(prepareKeyspace))) {
+                                    // This shouldn't happen in normal use, because a user shouldn't try to execute
+                                    // a prepared statement with the wrong keyspace set.
+                                    // Fail fast (we can't change the keyspace to reprepare, because we're using a pooled connection
+                                    // that's shared with other requests).
+                                    release(connection);
+                                    throw new IllegalStateException(String.format("Statement was prepared on keyspace %s, can't execute it on %s (%s)",
+                                        toPrepare.getQueryKeyspace(), connection.keyspace(), toPrepare.getQueryString()));
+                                }
+
+                                logger.info("Query {} is not prepared on {}, preparing before retrying executing. "
+                                        + "Seeing this message a few times is fine, but seeing it a lot may be source of performance problems",
+                                    toPrepare.getQueryString(), connection.address);
+
+                                write(connection, prepareAndRetry(toPrepare.getQueryString()));
+                                // we're done for now, the prepareAndRetry callback will handle the rest
+                                return;
+                            default:
+                                release(connection);
+                                if (metricsEnabled())
+                                    metrics().getErrorMetrics().getOthers().inc();
+                                break;
+                        }
+
+                        if (retry == null)
+                            setFinalResult(connection, response);
+                        else {
+                            switch (retry.getType()) {
+                                case RETRY:
+                                    ++retriesByPolicy;
+                                    if (logger.isDebugEnabled())
+                                        logger.debug("Doing retry {} for query {} at consistency {}", retriesByPolicy, statement, retry.getRetryConsistencyLevel());
+                                    if (metricsEnabled())
+                                        metrics().getErrorMetrics().getRetries().inc();
+                                    retry(true, retry.getRetryConsistencyLevel());
+                                    break;
+                                case RETHROW:
+                                    setFinalResult(connection, response);
```

Comment:
As far as RequestHandler is concerned, the goal is to get a definitive answer to the query. Whether that answer is a result or an exception makes no difference. The first execution that gives us this answer wins, and the others will be cancelled (the underlying assumption being that if an execution fails, the others would probably fail too).